File: tests/python/test_codecs.py
"""Optional VTK XML block codecs (zstd, lz4).

Guarded so the suite stays meaningful in a build with neither: the
availability, default-behaviour and error-message tests always run, and only
the round-trips are skipped.

The guarantee these exist to protect is the one that makes adding codecs safe
at all: **a pure build reads and writes exactly what it always did.** zlib
stays the default everywhere, and nothing about an existing file changes.
"""

from __future__ import annotations

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus.vtu import _vtu as _vtu_py

_core = pytest.importorskip("meshioplusplus._core")

HAS_ZLIB = getattr(_core, "__has_zlib__", False)
HAS_ZSTD = getattr(_core, "__has_zstd__", False)
HAS_LZ4 = getattr(_core, "__has_lz4__", False)

# The compressor= attribute each codec is recorded under. lz4 is a real VTK
# compressor; zstd is a meshio++ extension (VTK ships no ZSTD compressor).
ATTR = {
    "zlib": "vtkZLibDataCompressor",
    "lz4": "vtkLZ4DataCompressor",
    "zstd": "vtkZSTDDataCompressor",
}

# Unlike zstd/lz4 (default OFF), zlib defaults ON in CMake -- but it is still
# found via find_package, same as HDF5/netCDF, so a system without it (e.g.
# Windows CI, see CMakeLists.txt's MESHIOPLUSPLUS_WITH_ZLIB comment) compiles
# it out too. `meshioplusplus.write`/`.read` never notice (the Python fallback
# picks up zlib from the stdlib); only tests that call `_core` directly need
# this guard.
requires_zlib = pytest.mark.skipif(not HAS_ZLIB, reason="build has no zlib")
requires_zstd = pytest.mark.skipif(not HAS_ZSTD, reason="build has no zstd")
requires_lz4 = pytest.mark.skipif(not HAS_LZ4, reason="build has no lz4")


def _mesh():
    return meshioplusplus.Mesh(
        np.array([[0.0, 0.0, 0.0], [1.0, 0.0, 0.0], [1.0, 1.0, 0.0], [0.0, 1.0, 0.0]]),
        [("triangle", np.array([[0, 1, 2], [0, 2, 3]]))],
        point_data={"u": np.arange(4.0), "v": np.arange(4.0) * 3},
    )


def _assert_same(got, src):
    np.testing.assert_allclose(np.asarray(got.points)[:, :3], src.points)
    assert sorted(got.point_data) == sorted(src.point_data)
    for name, arr in src.point_data.items():
        np.testing.assert_allclose(np.asarray(got.point_data[name]), arr)


def test_zlib_capability_flag_exists():
    """Present as a boolean even when off (find_package can fail, e.g. Windows
    CI with no system zlib), matching the zstd/lz4 flags -- callers can branch
    without hasattr regardless of build configuration."""
    assert isinstance(getattr(_core, "__has_zlib__", None), bool)


def test_capability_flags_exist():
    # Present as booleans even when off, so callers can branch without hasattr.
    assert isinstance(_core.__has_zstd__, bool)
    assert isinstance(_core.__has_lz4__, bool)


@pytest.mark.parametrize("fmt", ["vtu", "vtp"])
def test_default_write_is_unchanged(tmp_path, fmt):
    """No codec asked for => zlib, exactly as before."""
    path = tmp_path / f"m.{fmt}"
    meshioplusplus.write(path, _mesh())
    text = path.read_text()
    assert ATTR["lz4"] not in text
    assert ATTR["zstd"] not in text
    _assert_same(meshioplusplus.read(path), _mesh())


@pytest.mark.parametrize(
    "codec",
    [
        "zlib",
        pytest.param("lz4", marks=requires_lz4),
        pytest.param("zstd", marks=requires_zstd),
    ],
)
@pytest.mark.parametrize("fmt", ["vtu", "vtp"])
def test_codec_round_trip(tmp_path, fmt, codec):
    path = tmp_path / f"m.{fmt}"
    src = _mesh()
    meshioplusplus.write(path, src, compression=codec)
    assert ATTR[codec] in path.read_text()
    _assert_same(meshioplusplus.read(path), src)


@pytest.mark.parametrize(
    "codec",
    [
        pytest.param("zlib", marks=requires_zlib),
        pytest.param("lz4", marks=requires_lz4),
        pytest.param("zstd", marks=requires_zstd),
    ],
)
def test_cpp_and_python_implementations_agree(tmp_path, codec):
    """Cross-compat both directions -- the real proof the framings match.

    A C++-written file must be readable by the pure-Python reference and vice
    versa; otherwise the two would silently be writing different formats under
    the same compressor= name.
    """
    src = _mesh()

    cpp_path = tmp_path / "cpp.vtu"
    _core.vtu_write_codec(str(cpp_path), src, True, codec)
    _assert_same(_vtu_py.read(cpp_path), src)

    py_path = tmp_path / "py.vtu"
    _vtu_py.write(py_path, src, binary=True, compression=codec)
    _assert_same(_core.vtu_read(str(py_path)), src)


def test_zlib_file_reads_in_a_build_without_optional_codecs(tmp_path):
    """A zlib file must never depend on zstd/lz4 being present."""
    path = tmp_path / "m.vtu"
    meshioplusplus.write(path, _mesh(), compression="zlib")
    assert ATTR["zlib"] in path.read_text()
    _assert_same(meshioplusplus.read(path), _mesh())


def test_unknown_codec_is_rejected(tmp_path):
    with pytest.raises(Exception, match="codec"):
        _core.vtu_write_codec(str(tmp_path / "m.vtu"), _mesh(), True, "bogus")


def _forge(tmp_path, codec):
    """A minimal VTU declaring `codec` without any real payload."""
    path = tmp_path / "forged.vtu"
    path.write_text(
        f'<VTKFile type="UnstructuredGrid" compressor="{ATTR[codec]}">'
        '<UnstructuredGrid><Piece NumberOfPoints="0"/>'
        "</UnstructuredGrid></VTKFile>"
    )
    return path


@pytest.mark.parametrize("codec", ["lz4", "zstd"])
def test_absent_codec_error_names_the_build_option(tmp_path, codec):
    """If the C++ core lacks a codec, the error must be actionable."""
    have = {"lz4": HAS_LZ4, "zstd": HAS_ZSTD}[codec]
    if have:
        pytest.skip(f"this build has {codec}")
    with pytest.raises(Exception) as excinfo:
        _core.vtu_read(str(_forge(tmp_path, codec)))
    assert "MESHIOPLUSPLUS_WITH_" in str(excinfo.value)


@pytest.mark.parametrize("codec", ["lz4", "zstd"])
def test_python_fallback_reports_the_missing_extra(codec):
    """With neither the C++ codec nor the Python module, fail by name.

    That combination is a genuinely new failure class -- a file readable by
    nothing in the stack -- so the message has to name the extra to install
    rather than surfacing an ImportError from deep inside the parser. Checked
    against the error path directly so it is exercised even when the modules
    *are* installed (as they are in CI).
    """
    shim = {"lz4": _vtu_py._lz4, "zstd": _vtu_py._zstd}[codec]
    message = str(shim._missing())
    assert "meshioplusplus[codecs]" in message
    assert f"MESHIOPLUSPLUS_WITH_{codec.upper()}=ON" in message
    assert {"lz4": "lz4", "zstd": "zstandard"}[codec] in message


def test_cli_compress_with_codec(tmp_path):
    from meshioplusplus._cli import main

    path = tmp_path / "m.vtu"
    meshioplusplus.write(path, _mesh())
    codec = "lz4" if HAS_LZ4 else "zlib"
    main(["compress", "--codec", codec, str(path)])
    assert ATTR[codec] in path.read_text()
    _assert_same(meshioplusplus.read(path), _mesh())


def test_cli_compress_rejects_codec_for_other_formats(tmp_path):
    """Silently ignoring --codec would be the worst outcome."""
    from meshioplusplus._cli import main

    path = tmp_path / "m.stl"
    meshioplusplus.write(path, _mesh())
    with pytest.raises(SystemExit):
        main(["compress", "--codec", "zstd", str(path)])


# ---------------------------------------------------------------------------
# Interop with a real VTK, which is the only thing that can actually settle it
# ---------------------------------------------------------------------------


def _vtk_grid_from(mesh):
    """Build the equivalent vtkUnstructuredGrid, for VTK-side writing."""
    import vtk

    points = vtk.vtkPoints()
    for p in mesh.points:
        points.InsertNextPoint(*p)
    grid = vtk.vtkUnstructuredGrid()
    grid.SetPoints(points)
    for block in mesh.cells:
        for row in block.data:
            cell = vtk.vtkTriangle()
            for i, node in enumerate(row):
                cell.GetPointIds().SetId(i, int(node))
            grid.InsertNextCell(cell.GetCellType(), cell.GetPointIds())
    return grid


def test_vtk_ships_lz4_but_not_zstd():
    """The fact that motivates treating the two codecs differently.

    lz4 is a real VTK compressor, so meshio++ writes it for interop. VTK has no
    ZSTD compressor, so `vtkZSTDDataCompressor` is explicitly a meshio++
    extension rather than a standard name we happen to support.
    """
    vtk = pytest.importorskip("vtk")
    assert hasattr(vtk, "vtkLZ4DataCompressor")
    assert not hasattr(vtk, "vtkZSTDDataCompressor")


@requires_lz4
def test_vtk_reads_our_lz4_output(tmp_path):
    """Our lz4 files must open in VTK/ParaView -- the whole point of using it."""
    vtk = pytest.importorskip("vtk")
    from vtk.util.numpy_support import vtk_to_numpy

    src = _mesh()
    path = tmp_path / "ours.vtu"
    _core.vtu_write_codec(str(path), src, True, "lz4")
    assert ATTR["lz4"] in path.read_text()

    reader = vtk.vtkXMLUnstructuredGridReader()
    reader.SetFileName(str(path))
    reader.Update()
    grid = reader.GetOutput()

    assert grid.GetNumberOfPoints() == len(src.points)
    np.testing.assert_allclose(vtk_to_numpy(grid.GetPoints().GetData()), src.points)


@requires_lz4
def test_we_read_vtks_lz4_output(tmp_path):
    """And the converse: VTK's own raw-block framing must decode here."""
    vtk = pytest.importorskip("vtk")

    src = _mesh()
    path = tmp_path / "fromvtk.vtu"
    writer = vtk.vtkXMLUnstructuredGridWriter()
    writer.SetFileName(str(path))
    writer.SetInputData(_vtk_grid_from(src))
    writer.SetDataModeToBinary()
    writer.SetCompressorTypeToLZ4()
    writer.Write()
    assert ATTR["lz4"] in path.read_text(errors="ignore")

    back = _core.vtu_read(str(path))
    np.testing.assert_allclose(np.asarray(back.points)[:, :3], src.points)


@requires_zstd
def test_vtk_cleanly_refuses_our_zstd_output(tmp_path):
    """zstd must fail *visibly* in VTK, never be silently misread.

    That is the trade being made by writing a non-VTK compressor name: ParaView
    reports an unknown compressor rather than producing wrong geometry.
    """
    vtk = pytest.importorskip("vtk")

    path = tmp_path / "ours.vtu"
    _core.vtu_write_codec(str(path), _mesh(), True, "zstd")

    reader = vtk.vtkXMLUnstructuredGridReader()
    reader.SetFileName(str(path))
    reader.SetGlobalWarningDisplay(0)  # the failure is expected; keep output clean
    reader.Update()
    # VTK cannot construct vtkZSTDDataCompressor, so no data comes through.
    assert reader.GetOutput().GetNumberOfPoints() == 0

File: tests/cpp/test_codecs.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//

// External includes
#include <gtest/gtest.h>

// System includes
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <limits>
#include <sstream>
#include <string>
#include <system_error>
#include <thread>
#include <vector>

// Project includes
#include "meshioplusplus/detail/vtk_cells.hpp"
#include "meshioplusplus/detail/vtu_binary.hpp"
#include "meshioplusplus/exceptions.hpp"
#include "meshioplusplus/formats/vtp.hpp"
#include "meshioplusplus/formats/vtu.hpp"
#include "mesh_fixtures.hpp"

using namespace meshioplusplus;
using meshioplusplus::detail::VtkCodec;

// The optional zstd/lz4 codecs. Everything here is guarded so the suite is
// meaningful in a pure build too: the availability and error-message tests run
// unconditionally, and only the round-trips are #ifdef'd.

namespace {

void codec_remove(const std::string& rPath) {
    std::error_code ec;
    std::filesystem::remove(rPath, ec);
}

std::string codec_slurp(const std::string& rPath) {
    std::ifstream in(rPath, std::ios::binary);
    std::ostringstream ss;
    ss << in.rdbuf();
    return ss.str();
}

}  // namespace

TEST(Codecs, ZlibIsAlwaysTheDefaultAndIsAvailable) {
    // The non-negotiable baseline: a build may lack zstd/lz4, but the default
    // write path must keep working.
    EXPECT_TRUE(detail::vtk_codec_available(VtkCodec::None));
#ifdef MESHIOPLUSPLUS_HAS_ZLIB
    EXPECT_TRUE(detail::vtk_codec_available(VtkCodec::Zlib));
#endif
}

TEST(Codecs, CompressorAttributeNames) {
    // LZ4's name is a real VTK class; ZSTD's is our documented extension.
    EXPECT_STREQ(detail::vtk_codec_compressor(VtkCodec::Zlib), "vtkZLibDataCompressor");
    EXPECT_STREQ(detail::vtk_codec_compressor(VtkCodec::LZ4), "vtkLZ4DataCompressor");
    EXPECT_STREQ(detail::vtk_codec_compressor(VtkCodec::ZSTD), "vtkZSTDDataCompressor");
    EXPECT_STREQ(detail::vtk_codec_compressor(VtkCodec::None), "");
}

TEST(Codecs, MissingCodecErrorNamesTheBuildOption) {
    // The whole point of the message: tell the user what to turn on.
    const std::string zstd_msg = detail::vtk_codec_missing_message(VtkCodec::ZSTD, false);
    EXPECT_NE(zstd_msg.find("MESHIOPLUSPLUS_WITH_ZSTD=ON"), std::string::npos) << zstd_msg;
    const std::string lz4_msg = detail::vtk_codec_missing_message(VtkCodec::LZ4, true);
    EXPECT_NE(lz4_msg.find("MESHIOPLUSPLUS_WITH_LZ4=ON"), std::string::npos) << lz4_msg;
    // lzma is recognized but never implemented, so it must not advertise a flag.
    const std::string lzma_msg = detail::vtk_codec_missing_message(VtkCodec::LZMA, false);
    EXPECT_NE(lzma_msg.find("not implemented"), std::string::npos) << lzma_msg;
}

TEST(Codecs, UnavailableCodecThrowsRatherThanFailingToLink) {
    // A codec compiled out must surface as a catchable error -- that is what
    // routes callers to the Python fallback instead of breaking the build.
    for (VtkCodec codec : {VtkCodec::Zlib, VtkCodec::LZ4, VtkCodec::ZSTD}) {
        if (detail::vtk_codec_available(codec))
            continue;
        EXPECT_THROW(detail::vtk_codec_require_read(codec), ReadError);
        EXPECT_THROW(detail::vtk_codec_require_write(codec), WriteError);
    }
}

// A pure build must keep reading what it always wrote. This is the guarantee
// that makes the codecs safe to add at all.
TEST(Codecs, ZlibFilesRoundTripRegardlessOfOptionalCodecs) {
#ifdef MESHIOPLUSPLUS_HAS_ZLIB
    const Mesh source = mt::data_mesh();
    const std::string path = mt::temp_path(".vtu");
    write_vtu(path, source, /*binary=*/true, /*zlib=*/true);

    const std::string text = codec_slurp(path);
    EXPECT_NE(text.find("vtkZLibDataCompressor"), std::string::npos);

    mt::expect_same_geometry(read_vtu(path), source);
    codec_remove(path);
#else
    GTEST_SKIP() << "build has no zlib";
#endif
}

namespace {

/** @brief Write with @p codec, read back, and compare against the source. */
void codec_expect_roundtrip(VtkCodec codec, const char* pCompressor) {
    const Mesh source = mt::data_mesh();
    const std::string path = mt::temp_path(".vtu");
    write_vtu_codec(path, source, /*binary=*/true, codec);

    const std::string text = codec_slurp(path);
    EXPECT_NE(text.find(pCompressor), std::string::npos) << "compressor attribute not recorded";

    const Mesh back = read_vtu(path);
    mt::expect_same_geometry(back, source);
    EXPECT_EQ(back.PointDataNames(), source.PointDataNames());
    EXPECT_EQ(back.CellDataNames(), source.CellDataNames());

    codec_remove(path);
}

}  // namespace

TEST(Codecs, ZstdRoundTrip) {
#ifdef MESHIOPLUSPLUS_HAS_ZSTD
    codec_expect_roundtrip(VtkCodec::ZSTD, "vtkZSTDDataCompressor");
#else
    GTEST_SKIP() << "build has no zstd";
#endif
}

TEST(Codecs, Lz4RoundTrip) {
#ifdef MESHIOPLUSPLUS_HAS_LZ4
    codec_expect_roundtrip(VtkCodec::LZ4, "vtkLZ4DataCompressor");
#else
    GTEST_SKIP() << "build has no lz4";
#endif
}

TEST(Codecs, VtpRoundTripsWithOptionalCodecs) {
    const Mesh source = mt::tri_mesh();
#ifdef MESHIOPLUSPLUS_HAS_LZ4
    {
        const std::string path = mt::temp_path(".vtp");
        write_vtp_codec(path, source, /*binary=*/true, VtkCodec::LZ4);
        mt::expect_same_geometry(read_vtp(path), source);
        codec_remove(path);
    }
#endif
#ifdef MESHIOPLUSPLUS_HAS_ZSTD
    {
        const std::string path = mt::temp_path(".vtp");
        write_vtp_codec(path, source, /*binary=*/true, VtkCodec::ZSTD);
        mt::expect_same_geometry(read_vtp(path), source);
        codec_remove(path);
    }
#endif
    SUCCEED();
}

// Compression must change only the framing. Every codec's payload has to match
// the uncompressed variant's exactly, which is what proves the block framing is
// shared rather than reimplemented per codec.
TEST(Codecs, EveryCodecDecodesToTheSameBytes) {
    const std::string data(100000, '\0');
    std::string payload = data;
    for (std::size_t i = 0; i < payload.size(); ++i)
        payload[i] = static_cast<char>(i * 31 + (i >> 3));
    const unsigned char* bytes = reinterpret_cast<const unsigned char*>(payload.data());

    const std::string raw = detail::vtu_encode_binary(bytes, payload.size(), VtkCodec::None);
    const std::vector<unsigned char> expected =
        detail::vtu_decode_uncompressed(raw.c_str(), raw.size(), 4);
    ASSERT_EQ(expected.size(), payload.size());

    for (VtkCodec codec : {VtkCodec::Zlib, VtkCodec::LZ4, VtkCodec::ZSTD}) {
        if (!detail::vtk_codec_available(codec))
            continue;
        const std::string encoded = detail::vtu_encode_binary(bytes, payload.size(), codec);
        const std::vector<unsigned char> decoded =
            detail::vtu_decode_blocks(encoded.c_str(), encoded.size(), 4, codec);
        EXPECT_EQ(decoded, expected)
            << "codec " << detail::vtk_codec_name(codec) << " changed the payload";
        // Multi-block: 100000 bytes exceeds the fixed 32 KiB block size.
        EXPECT_LT(encoded.size(), raw.size())
            << "codec " << detail::vtk_codec_name(codec) << " did not compress";
    }
}

TEST(Codecs, UInt64HeadersRoundTripAndUInt32KeepsItsBytes) {
    std::string payload(70000, '\0');
    for (std::size_t i = 0; i < payload.size(); ++i)
        payload[i] = static_cast<char>(i * 17 + (i >> 5));
    const unsigned char* bytes = reinterpret_cast<const unsigned char*>(payload.data());
    const std::vector<unsigned char> expected(bytes, bytes + payload.size());

    for (VtkCodec codec : {VtkCodec::None, VtkCodec::Zlib}) {
        if (codec != VtkCodec::None && !detail::vtk_codec_available(codec))
            continue;
        // header_type="UInt64": every size in the header is 8 bytes wide.
        const std::string wide = detail::vtu_encode_binary(bytes, payload.size(), codec, 8);
        const std::vector<unsigned char> decoded =
            codec == VtkCodec::None
                ? detail::vtu_decode_uncompressed(wide.c_str(), wide.size(), 8)
                : detail::vtu_decode_blocks(wide.c_str(), wide.size(), 8, codec);
        EXPECT_EQ(decoded, expected) << detail::vtk_codec_name(codec);
        // The UInt32 overload writes exactly what the three-argument form did.
        EXPECT_EQ(detail::vtu_encode_binary(bytes, payload.size(), codec, 4),
                  detail::vtu_encode_binary(bytes, payload.size(), codec))
            << detail::vtk_codec_name(codec);
    }
}

TEST(Codecs, HeaderTypeWidensOnlyPastThirtyTwoBits) {
    constexpr std::uint64_t kMax32 = std::numeric_limits<std::uint32_t>::max();
    EXPECT_EQ(detail::vtu_header_bytes_for(0), 4u);
    EXPECT_EQ(detail::vtu_header_bytes_for(kMax32), 4u);
    EXPECT_EQ(detail::vtu_header_bytes_for(kMax32 + 1), 8u);
    // A small mesh keeps UInt32, so its files keep their bytes.
    EXPECT_EQ(detail::vtk_xml_header_bytes(mt::tet_mesh()), 4u);

    // An uncompressed array of 4 GiB under a UInt32 header is refused, not
    // truncated -- checked before the payload is read or allocated.
    const unsigned char byte = 0;
    EXPECT_THROW(detail::vtu_encode_binary(&byte, std::size_t{kMax32} + 1, VtkCodec::None, 4),
                 WriteError);
    EXPECT_THROW(detail::vtu_encode_binary(&byte, 1, VtkCodec::None, 2), WriteError);
}

TEST(Codecs, Base64DecodesIdenticallyFromManyThreads) {
    std::string payload(50000, '\0');
    for (std::size_t i = 0; i < payload.size(); ++i)
        payload[i] = static_cast<char>(i * 7 + 3);
    const std::string text =
        detail::b64encode(reinterpret_cast<const unsigned char*>(payload.data()), payload.size());
    // Its inverse table is built on first use, once and thread-safely: calls
    // racing to be first must agree.
    std::vector<std::vector<unsigned char>> results(8);
    std::vector<std::thread> threads;
    for (std::size_t t = 0; t < results.size(); ++t)
        threads.emplace_back([&, t] { results[t] = detail::b64decode(text.c_str(), text.size()); });
    for (auto& th : threads)
        th.join();
    const std::vector<unsigned char> expected(payload.begin(), payload.end());
    for (const auto& r : results)
        EXPECT_EQ(r, expected);
}

TEST(Codecs, ReadingAFileNeedingAnAbsentCodecReportsTheOption) {
    // Hand-built VTU declaring a codec; if this build lacks it, the reader must
    // say which option to enable rather than failing obscurely.
    for (VtkCodec codec : {VtkCodec::LZ4, VtkCodec::ZSTD}) {
        if (detail::vtk_codec_available(codec))
            continue;
        const std::string path = mt::temp_path(".vtu");
        {
            std::ofstream os(path);
            os << "<VTKFile type=\"UnstructuredGrid\" compressor=\""
               << detail::vtk_codec_compressor(codec) << "\"><UnstructuredGrid>"
               << "<Piece NumberOfPoints=\"0\"/></UnstructuredGrid></VTKFile>";
        }
        try {
            read_vtu(path);
            ADD_FAILURE() << "expected a ReadError for " << detail::vtk_codec_name(codec);
        } catch (const ReadError& e) {
            EXPECT_NE(std::string(e.what()).find("MESHIOPLUSPLUS_WITH_"), std::string::npos)
                << e.what();
        }
        codec_remove(path);
    }
    SUCCEED();
}
